Initial commit for generating relaxations from a cat file.

hugookeeffe opened the pull request:
This commit introduces a new tool to generate relaxations from a cat file; this is still a work in progress. Any and all feedback would be appreciated!

The tool takes as input a cat file and then tries to translate each let statement into a list of relaxations, where the user can choose which let statements to output.

In order to use the tool, a user can use the command `mcat2config7` and must then specify a cat file. 
With the option `-tree`, the cat file let statements will all be expanded and this will be represented. 
Using the option `-let` the user can choose what let statements to print relaxations for.
e.g. `mcat2compare7 -tree -let ob`

The tool works as follows:
- Parse AST of cat file
- Expand Unions within a let statement
- If a let statement has a variable that corresponds to a let statement, implement that and expand
- translate variables from cat to type edge (from edge.ml)
- Match cumulative relaxations
- Print

Currently, in the stage where variables are translated to edges, there are still many edges unimplemented. 
In the expansion of the cat file logic, some instructions are also not yet implemented, i.e. Diff, add, Inv, Opt etc.

I will be implementing these in the following patches, however I am making a draft PR in order to get some feedback on the general logic of the tool before implementing everything.
I have tried to make my functions easy to understand, however this might not have succeeded in all places. If there's anything that doesn't make sense then I can clarify or change it to be more understandable.

any feedback would be appreciated

HadrienRenaud commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,666 @@
+        | [] -> raise (Failed "cannot provide empty list")
+      in f expl empty_node
+    
+    and listofnode nodel = 
```
```suggestion
    (** [listofnode nodel] returns the list of possible paths starting in [nodel]. *)
    and listofnode nodel = 
```

HadrienRenaud reviewed (commented):
I tried to review but it's often a bit too complicated for me, I would need far more comments or time to understand precisely what is going on. Maybe it's my lack of familiarity with the `gen` code.

In any case, I have commented a few things, feel free **not** to follow them. Even in some case I have commented a few comments that helped my understanding of a few functions.

hugookeeffe commented:
The newest patch adds support for rmw instructions and the range function into the non-refactored version of the code. This allows the tool to generate relaxations for the entirety of aarch64-v00.cat. 

This patch does not yet address the comments made by Hadrien; I am working on a refactored patch which I will upload soon that addresses the comments

hugookeeffe commented:
@HadrienRenaud I think I have responded to all of your comments. The biggest changes were refactoring the code with the new type for expr and ast. I have also tried to add as many comments as possible for the major functions, if anything is still unclear let me know and I can elaborate.

I have also separately added support for rmw instructions

HadrienRenaud reviewed (commented):
Hi @hugookeeffe,

I can't build it on my machine. If you need any help with solving the warnings, I can help with that.
Maybe the unknown errors that you have to catch with `try .. with` come from the non-exhaustive pattern-matching that you have. This is actually warned by ocaml with warning n.8 which you should not silence but resolve.

Most of the comments I make are towards better understanding/simplification of the code. However, there are a few points where I don't understand what is happening and I'd be grateful if you could help me with that.

For warnings 40 and 42, imho, you could just ignore them and add a line at the begining like:
```ocaml
[@@@warning "-40-42"]
```
For the other warnings, imho, you should fix them, they are quite important.

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,666 @@
+            begin match br.vartype with 
+            | Var -> begin
+              let vtype = 
+                try match_var br with
```
I have included some more comments, if you are happy with them I can resolve this

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,730 @@
+      end
+      | _ -> raise (Failed "wrong inter")
+   
+    and unroll_inter (Op2 (a, exp)) = 
```
It wasn't working properly, I have fixed it to make sure unroll_inter only works in the correct case

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,730 @@
+      Output: 
+      - A list of lists of expressions representing all possible expansions of the input list.
+      *)
+      match input_item with
+      | [] -> [[]]
+      | hd::tl ->
+          let expanded_hd = expand_item hd in
+          let expanded_tl = expand_list tl in
+          List.concat (List.map (fun hd_item ->
+            List.map (fun tl_item ->
+              hd_item @ tl_item
+            ) expanded_tl
+          ) expanded_hd)
```
I haven't been able to get this working. any help with getting this to work would be appreciated

hugookeeffe commented:
Hi @HadrienRenaud 

Thanks for the comments! I have resolved all of the questions I think I have answered, if you don't think they should be resolved, let me know and I can look at it again.

Many thanks with all the help :)

relokin reviewed (commented):
Some initial thoughts.

There are still some basic cases that won't work. For example I tried:
```
mcat2config7 -let ca catalogue/aarch64/cats/aarch64-v00.cat
```
and I didn't get anything as an output.

From a design point of view, I think we could optimize the tool to expand and translate from the cat language (events, relations and operators) to relaxations only when needed - think lazy evaluation. For example, in the above example:
```
mcat2config7 -let ca catalogue/aarch64/cats/aarch64-v00.cat
```
We transform the whole cat file into relaxations and then the tool looks for the definition of ca. I think it would easier for the tool for find the definition of ca
```
let ca = co | fr
```
and then use match_var to map it to relations
co -> Coe Coi
fr -> Fre Fri
and then expand the union operator.

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,729 @@
+    
+    type ast = ( string * expr) list
+
+    let empty_node = ["",Empty ""]
```
removed

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,729 @@
+      | Dd of extr*(atom option) * string
+      | Lrs of string
+    
+    type ast = ( string * expr) list
```
I have now changed the code slightly to explicitly use this type. I am going to further update the typing used in the next patch

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,729 @@
+      | Let (_, (_,Pvar (Some varname),expression) :: _) -> varname,expression
+      | _ -> raise (Failed "instruction not supported\n" )
+
+    and get_tags tree (varname,expression)  = 
```
you are right, it is not necessary and I have removed it. I had to make a slight modification when calling get_vars, to return the name of the let statement and the expression in a tuple, as can be seen in line 542

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,729 @@
+          | _ -> (tree f t)
+        end
+        | [] -> [] in
+      let tree_base = tree get_ins ast in
```
that's correct, I hadn't yet added fr and co to my matching function, this should now be done. Thanks for noticing! I have also attempted to implement some lazy evaluation like you suggested, this can be seen in the map_vars function on line 539. Please let me know if this is not what you intended :)

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,729 @@
+    ------------------------------------------   
+    *)
+    and get_tree ast = 
+      let rec tree f l = 
```
I have added it. I was struggling to use fold_left previously as I wanted to be able to continue the fold after a fail, but I think I have implemented this now.

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,729 @@
+       ---------------------------------------------------------------
+       Entry point
+       --------------------------------------------------------------- *)
+    let rec get_ast fname =
```
done

hugookeeffe commented on gen/mcat2config.ml:
I have it in gen as I use aarch64gen.ml and edge.ml. If its possible to do this with my tool in tools/ I am happy to do this, however I couldn't find a way to do this previously.

hugookeeffe commented:
thank you for your comments @relokin ! Hopefully I have responded to all of them now. 
There are some things I am going to implement in my next patch, namely:
- improve my internal representation for ast, by splitting it into multiple types. This would ideally help with making the code easier to understand (suggested by @HadrienRenaud )
- implement new exception type to represent unimplemented features, having everything under Failed makes it confusing (suggested by @relokin and @HadrienRenaud )

hugookeeffe commented:
I have just pushed a patch implementing the new type system, as suggested by @HadrienRenaud and the new exception type as suggested by @relokin and @HadrienRenaud 

I have tried to clean up the code as well to make it more understandable with the new type system. As I refactored a lot of the code there might be some minor errors, so if you see anything please let me know.

Feedback on whether the code is easier to understand now would be very much appreciated.

HadrienRenaud reviewed (commented):
Hi @hugookeeffe,

Thank you for the update! Everything looks far more explicit, I like that.

However, I think you could still go a bit further in that direction, getting even more explicit. A lot of my comments points to that, They are suggestions, feel free to not follow them. There are a lot of them but they all point towards not having a single function returning or taking a list as an argument 😈 

What is not avoidable I'm afraid is the `List.concat_map` thing: we need herdtools and your tool with it to compile on `4.08.1`, see my comment on that. (Small tip: install an opam switch somewhere  and try to compile your branch on it! maybe on a server or something)

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -204,149 +159,173 @@ module Make
+          Outputs: list of let statements with var replaced by edge_ir in the ir type
+        *)
+        let edges_of_sequence expl =
+          List.concat_map
```
thanks for letting me know! I have made this change now

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -204,149 +159,173 @@ module Make
+            - tree: list of let statemenets
+          Outputs: list of let statements with var replaced by edge_ir in the ir type
+        *)
+        let edges_of_sequence expl =
```
ah yes, that makes a lot of sense. I have implemented this change

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,703 @@
+let verbose = ref 0
+let lets_to_print = ref []
+let arg = ref []
+let print_tree = ref false
```
thanks for the suggestion! I have implemented this

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,703 @@
+          try
+            f a::acc
+          with
+            | NotImplemented s -> if O.verbose > 0 then (eprintf "%s\n" s);
```
normally it doesn't I think. all unexpected exceptions should raise a fatal error which is only caught at the top level, so the whole code will stop running

hugookeeffe commented on gen/mcat2config.ml:
```diff
@@ -0,0 +1,703 @@
+          fun (varname, expression) ->
+          match List.find_opt (fun name -> String.equal name varname) O.lets_to_print with
+          | Some _ -> varname, inline_vars varname tree_base expression
+          | None -> raise (NotImplemented (sprintf "not generating for instruction: %s\n" varname))
```
I think you're right; I have also changed this to a fatal exception anyway, as the program should stop if there are no let statements that match what the user asked for

hugookeeffe commented:
Hi @maranget, if you have time, could you review this PR? This PR implements a tool to generate relaxations from a cat file.
I have just moved it out of draft. There is quite a lot of conversation and commit history, so if you want me to clarify / simplify anything please let me know.

hugookeeffe commented:
The newest commit forces the tool to always fail if it comes across a function or keyword that it doesn't know. This is to prevent the generation of incomplete or wrong relaxations
